contracts/RealmDiamond/facets/ParcelRolesRegistryFacet.sol

```diff
@@ -30,16 +30,22 @@ contract ParcelRolesRegistryFacet is Modifiers, IERC7432 {
   function grantRole(Role calldata _role) external override onlyValidRole(_role.roleId) onlyRealm(_role.tokenAddress) {
     require(_role.expirationDate > block.timestamp, "ParcelRolesRegistryFacet: expiration date must be in the future");
 
-    // Deposit NFT if necessary and get the original owner
     address _originalOwner = _depositNft(_role.tokenAddress, _role.tokenId);
 
-    // Get role data from storage
     RoleData storage _roleData = s.erc7432_roles[_role.tokenAddress][_role.tokenId][_role.roleId];
 
-    // Check if the role is either revocable or expired
     require(_roleData.revocable || _roleData.expirationDate < block.timestamp, "ParcelRolesRegistryFacet: role must be expired or revocable");
 
-    s.erc7432_roles[_role.tokenAddress][_role.tokenId][_role.roleId] = RoleData(_role.recipient, _role.expirationDate, _role.revocable, _role.data);
+    if (_role.data.length > 0) {
+      if (_role.roleId == keccak256("AlchemicaChanneling()") || _role.roleId == keccak256("EmptyReservoir()")) {
+        _handleProfitShareData(_role.tokenAddress, _role.tokenId, _role.roleId, _role.data);
+      }
+    }
```

Comment:
First check if the supports profit share (by checking if it's alchemica or empty reservoir). If supports profit share, validate & store decoded profit share, if it does not support it, validate that `_role.data` is empty